### PR TITLE
feat/milesight-connectors/APPD-391

### PR DESCRIPTION
**Pull request by mateuscardosodeveloper**
## Decoder Description

Added new decoders for Milesight sensors.

## Type of change

- [x] Adding a new Decoder of Connector
- [ ] Adding a new Decoder of Network
- [ ] Update or fixing an issue in existing Decoder

## Decoder Information and Payload to test and review

- [x] Documentation of the hardware or protocol:
- [x] Payload of example the test the decoder:

## Checklist for Adding a New Decoder

- [x] Created a new folder under `./decoders/network/` or `./decoders/connector/` with the name of your decoder.
- [x] Added a `network.jsonc` or `connector.jsonc` file that follows the structure defined in `./schema/`.
- [x] Created version folders and added `manifest.jsonc` files for each version.
- [x] Followed the folder structure guidelines for manufacturer and sensor/device model.
- [x] The code has unit test and it's in TypeScript.

### CODE REVIEW

File: decoders/connector/milesight/ct101/description.md [hunk 1]
```diff
@@ -0,0 +1 @@
+The Milesight CT101 LoRaWAN Sensor is a unique current monitoring solution designed for seamless integration into existing electrical systems. This self-powered sensor eliminates the need for batteries, offering long-term operation and reduced maintenance.
```
Comment (Freddyminu) [inline decoders/connector/milesight/ct101/description.md:1]: Make this smaller, don't forget that this shows up here:

image

File: decoders/connector/milesight/ct101/v1.0.0/payload-config.jsonc [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  "description": "../description.md",
+  "install_text": "The Milesight CT101 LoRaWAN Sensor is a unique current monitoring solution designed for seamless integration into existing electrical systems. This self-powered sensor eliminates the need for batteries, offering long-term operation and reduced maintenance.\n\n**Key Features:**\n* **Self-powered:** Derives power directly from the measured circuit, eliminating battery dependence and ensuring continuous operation.\n* **Multiple current options:** Supports various current ranges to cater to diverse monitoring needs.\n* **High accuracy:** Delivers exceptional accuracy with a high sampling frequency of up to 3.3 kHz.\n* **Compact and easy installation:** Features a compact design and clamp-on functionality for quick and safe installation without interrupting the circuit.\n* **LoRaWAN connectivity:** Enables long-range, low-power communication for efficient data transmission.\n* **Threshold alarms:** Allows for setting thresholds to trigger alerts when current levels exceed or fall below predefined values.\n* **Remote monitoring**: Access real-time and historical data remotely through a LoRaWAN gateway and cloud platform.",
```
Comment (Freddyminu) [inline decoders/connector/milesight/ct101/v1.0.0/payload-config.jsonc:4]: When creating the install_text make sure the formatting is correct, talking in regards to the \n * etc...

**Review by Freddyminu — CHANGES_REQUESTED** [review]
Make the subtitle (description.md) smaller and concise. Do not utilize any subjective words, utilize only facts when describing the device.  

Utilize metadata on the variables when deemed useful. Create variable names that are intuitive and easy to understand.

**Review by Freddyminu — CHANGES_REQUESTED** [review]

File: decoders/connector/milesight/ct101/v1.0.0/payload-config.jsonc [hunk 3]
```diff
@@ -0,0 +1,25 @@
+{
+  "$schema": "../../../../../schema/connector_details.json",
+  "description": "../description.md",
+  "install_text": "The **CT101** is a LoRaWAN® Smart Current Transformer for remote energy monitoring and consumption analysis. It offers multiple current options and supports threshold alarms. The device features a detachable, compact, and clamp design for quick and safe indoor installation without de-energizing facilities. Compatible with Milesight LoRaWAN® gateway and Milesight IoT Cloud, it allows remote monitoring via webpage or mobile app. Suitable for energy monitoring in smart buildings and machine failure detection.",
```
Comment (Freddyminu) [inline decoders/connector/milesight/ct101/v1.0.0/payload-config.jsonc:4]: Here.

File: decoders/connector/milesight/ct305-ct310/v1.0.0/payload-config.jsonc [hunk 4]
```diff
@@ -0,0 +1,25 @@
+{
+  "$schema": "../../../../../schema/connector_details.json",
+  "description": "../description.md",
+  "install_text": "The **CT305/CT310** is a LoRaWAN® Smart Current Transformer designed for monitoring energy consumption and analyzing usage remotely. It reports RMS current and accumulated current data by minutes with a high measuring accuracy and a sampling frequency of up to 3.3 kHz. The device is self-powered and features a non-invasive clamp design for easy installation without de-energizing facilities. It supports a sampling rate of up to 1s for real-time monitoring and quick alarm response.\n The CT305/CT310 can detect three phases with a detection range of either 500A or 1000A. It includes an LED indicator for status and alarms and supports an external wire temperature sensor. The device is compliant with standard LoRaWAN® gateways and the Milesight Development Platform for remote monitoring.",
```
Comment (Freddyminu) [inline decoders/connector/milesight/ct305-ct310/v1.0.0/payload-config.jsonc:4]: Here.

File: decoders/connector/milesight/em400-mud-lora/description.md [hunk 5]
```diff
@@ -0,0 +1 @@
+Ultrasonic distance sensor with 3-axis accelerometer, temperature sensor, LoRaWAN
```
Comment (Freddyminu) [inline decoders/connector/milesight/em400-mud-lora/description.md:1]: How about "Ultrasonic distance sensor with a 3-axis accelerometer, temperature sensor, **_over_** LoRaWAN."

File: decoders/connector/milesight/em400-mud-lora/v1.0.0/payload-config.jsonc [hunk 6]
```diff
@@ -0,0 +1,25 @@
+{
+  "$schema": "../../../../../schema/connector_details.json",
+  "description": "../description.md",
+  "install_text": "**EM400-MUD** is a multifunctional ultrasonic distance sensor designed for small-range detection and minimal blind spots. It includes a 3-axis accelerometer and a temperature sensor for device status monitoring. The sensor features a high IP rating and a waterproof enclosure, capable of operating in harsh environments for up to 10 years with two 9000 mAh batteries. Available in LoRaWAN®  versions, it supports various communication needs. The LoRaWAN® version integrates with Milesight LoRaWAN® gateway and IoT Cloud.",
```
Comment (Freddyminu) [inline decoders/connector/milesight/em400-mud-lora/v1.0.0/payload-config.jsonc:4]: Here.

File: decoders/connector/milesight/em400-mud-nb-iot/description.md [hunk 7]
```diff
@@ -0,0 +1 @@
+Ultrasonic distance sensor with 3-axis accelerometer, temperature sensor, NB-IoT
```
Comment (Freddyminu) [inline decoders/connector/milesight/em400-mud-nb-iot/description.md:1]: Add **_over_** before NB-IoT

File: decoders/connector/milesight/em400-mud-nb-iot/v1.0.0/payload.ts [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+    // DISTANCE
+    else if (channel_id === 0x04 && channel_type === 0x82) {
+      decoded.distance = buffer.readUInt16LE();
+    }
+    // POSITION
+    else if (channel_id === 0x05 && channel_type === 0x00) {
+      decoded.position = buffer.readUInt8() === 0 ? "normal" : "tilt";
+    }
+    // LOCATION
+    else if (channel_id === 0x06 && channel_type === 0x88) {
```
Comment (Freddyminu) [inline decoders/connector/milesight/em400-mud-nb-iot/v1.0.0/payload.ts:76]: When handling location data you must create the "location" variable. It must follow the following data structure example:
 `{
    "variable": "location",
    "value": "-23.550556,-46.633146",
    "location": {
      "lat": -23.550556,
      "lng": -46.633146
    }
  }`

File: decoders/connector/milesight/em400-udl-lora/v1.0.0/payload-config.jsonc [hunk 9]
```diff
@@ -0,0 +1,25 @@
+{
+  "$schema": "../../../../../schema/connector_details.json",
+  "description": "../description.md",
+  "install_text": "The **EM400-UDL** is a non-contact ultrasonic distance sensor designed for outdoor applications such as monitoring water levels, fill levels of tanks and silos, object presence, or snow levels. It features multiple selective probes with varying detection ranges. The sensor has a high protection IP rating and a waterproof enclosure, allowing it to withstand harsh environments. It operates for up to 10 years using two 9000 mAh batteries. The LoRaWAN® version integrates with Milesight LoRaWAN® gateway and IoT Cloud solution.",
```
Comment (Freddyminu) [inline decoders/connector/milesight/em400-udl-lora/v1.0.0/payload-config.jsonc:4]: Here.

File: decoders/connector/milesight/em400-udl-nb-iot/v1.0.0/payload.ts [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+    else if (channel_id === 0x04 && channel_type === 0x82) {
+      decoded.distance = buffer.readUInt16LE();
+    }
+    // POSITION
+    else if (channel_id === 0x05 && channel_type === 0x00) {
+      decoded.position = buffer.readUInt8() === 0 ? "normal" : "tilt";
+    }
+    // LOCATION
+    else if (channel_id === 0x06 && channel_type === 0x88) {
+      decoded.latitude = buffer.readInt32LE() / 1000000;
```
Comment (Freddyminu) [inline decoders/connector/milesight/em400-udl-nb-iot/v1.0.0/payload.ts:77]: When handling location data you must create the "location" variable. It must follow the following data structure example:
 `{
    "variable": "location",
    "value": "-23.550556,-46.633146",
    "location": {
      "lat": -23.550556,
      "lng": -46.633146
    }
  }`

File: decoders/connector/milesight/vs330/description.md [hunk 11]
```diff
@@ -0,0 +1 @@
+LoRaWAN® bathroom occupancy sensor with ToF and PIR, over 99.5% accuracy, low power.
```
Comment (Freddyminu) [inline decoders/connector/milesight/vs330/description.md:1]: Remove the accuracy rating.

File: decoders/connector/milesight/vs340/v1.0.0/payload-config.jsonc [hunk 12]
```diff
@@ -0,0 +1,25 @@
+{
+  "$schema": "../../../../../schema/connector_details.json",
+  "description": "../description.md",
+  "install_text": "Milesight **VS340** is an occupancy sensor designed to detect desk or seat occupancy in workspaces using PIR technology. The VS341 Pro version includes additional Thermopile IR technology for enhanced accuracy. The sensor offers up to 98% accuracy, supports different PIR covers for adjustable field angles and detection ranges, and utilizes the Milesight D2D protocol for low-latency control without gateways. It is made from V0 grade flame retardant material and features NFC for one-touch configuration. The sensor is compatible with standard LoRaWAN® gateways, network servers, and Milesight IoT Cloud.",
```
Comment (Freddyminu) [inline decoders/connector/milesight/vs340/v1.0.0/payload-config.jsonc:4]: Here.

File: decoders/connector/milesight/vs341/v1.0.0/payload-config.jsonc [hunk 13]
```diff
@@ -0,0 +1,25 @@
+{
+  "$schema": "../../../../../schema/connector_details.json",
+  "description": "../description.md",
+  "install_text": "Milesight **VS341** is an occupancy sensor designed to detect desk or seat occupancy in workspaces. It uses a high-precision PIR sensor and Thermopile IR technology to achieve up to 98% accuracy. The sensor supports the Milesight D2D protocol for ultra-low latency communication and is compatible with standard LoRaWAN® gateways and network servers. It features NFC for one-touch configuration and is made from V0 grade flame retardant material for enhanced safety. The sensor is also compatible with Milesight IoT Cloud",
```
Comment (Freddyminu) [inline decoders/connector/milesight/vs341/v1.0.0/payload-config.jsonc:4]: Here.

File: decoders/connector/milesight/ws202/v1.0.0/payload-config.jsonc [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
...r occupancy. The WS202 can detect whether there is movement within a range of 6-8 meters and send the changes via the LoRaWAN® network. Additionally, the WS202 is equipped with a light sensor that can link PIR detection results to trigger scenes. With its compact size and battery power, the WS202 is easy to install anywhere. \n\nCompliant with the Milesight LoRaWAN® gateway and Milesight IoT Cloud solution, users can monitor alarms or room occupancy and easily trigger other sensors or appliances remotely via a webpage or mobile app.\n\nThe WS202 can be widely used in smart homes, smart offices, schools, warehouses, and more.\n\n**Features**\n* Occupancy or motion detection based on passive infrared and Fresnel Lens\n* Easily to install with the convenient size\n* Built-in light sensor, combine PIR sensor to achieve triggers\n* Ultra-wide-distance wireless transmission up to line of sight of 15 km\n* Equipped with NFC for one touch configuration, support card emulation mode\n* Support Milesight LoRa D2D protocol to enable ultra-low latency and directly control without gateway\n* Function well with standard LoRaWAN® gateways and network servers\n* Compliant with Milesight IoT Cloud",
```
Comment (Freddyminu) [inline decoders/connector/milesight/ws202/v1.0.0/payload-config.jsonc:4]: Here.

Comment (Freddyminu) [inline decoders/connector/milesight/ws202/v1.0.0/payload-config.jsonc:4]: Remove subjective information: "users can monitor alarms or room occupancy and **_easily_** trigger other sensors or appliances remotely via a webpage or mobile app."

File: decoders/connector/milesight/ws203/v1.0.0/payload-config.jsonc [hunk 15]
```diff
@@ -0,0 +1,25 @@
+{
+  "$schema": "../../../../../schema/connector_details.json",
+  "description": "../description.md",
+  "install_text": "**Milesight WS203** is an environmental monitoring device that combines motion detection using passive infrared (PIR) technology and temperature/humidity (TH) sensors. It detects motion within an 8-meter range and triggers TH detection, reporting environmental status periodically over the LoRaWAN® network. The device supports NFC for configuration and can store up to 1000 historical records with retransmission capability. It is compatible with standard LoRaWAN® gateways, Milesight IoT Cloud, and supports the Milesight D2D protocol for direct control without a gateway.",
```
Comment (Freddyminu) [inline decoders/connector/milesight/ws203/v1.0.0/payload-config.jsonc:4]: Here.

File: decoders/connector/milesight/ws302/v1.0.0/payload-config.jsonc [hunk 16]
```diff
@@ -0,0 +1,25 @@
+{
+  "$schema": "../../../../../schema/connector_details.json",
+  "description": "../description.md",
+  "install_text": "Milesight **WS302** is a LoRa sound level sensor with an integrated microphone, capable of measuring a wide range of noise levels. It supports multiple weighting measurements and transmits data via the LoRaWAN® network. The sensor is compatible with Milesight LoRaWAN® gateway and Milesight IoT Cloud solution for remote management and integration with other sensors or appliances.",
```
Comment (Freddyminu) [inline decoders/connector/milesight/ws302/v1.0.0/payload-config.jsonc:4]: Here.

File: decoders/connector/milesight/ws523/v1.0.0/payload-config.jsonc [hunk 17]
```diff
@@ -0,0 +1,25 @@
+{
+  "$schema": "../../../../../schema/connector_details.json",
+  "description": "../description.md",
+  "install_text": "Milesight **WS523** is a LoRaWAN® smart portable socket designed for monitoring and controlling electrical appliances. It supports various international socket types and allows remote or local switching of devices.\nThe WS523 measures current, voltage, power, and electrical consumption, and provides overcurrent alarms. It features an ultra-wide-distance wireless transmission up to 15 km line of sight and is equipped with NFC for one-touch configuration. The device is compatible with standard LoRaWAN® gateways, network servers, and the Milesight IoT Cloud solution.",
```
Comment (Freddyminu) [inline decoders/connector/milesight/ws523/v1.0.0/payload-config.jsonc:4]: Here.

File: decoders/connector/milesight/wt201/description.md [hunk 18]
```diff
@@ -0,0 +1 @@
+LoRaWAN® thermostat for HVAC systems, real-time monitoring, remote management, 4.2” LCD.
```
Comment (Freddyminu) [inline decoders/connector/milesight/wt201/description.md:1]: remove screen size, add "LCD Panel"

File: decoders/connector/milesight/wt201/v1.0.0/payload-config.jsonc [hunk 19]
```diff
@@ -0,0 +1,25 @@
+{
+  "$schema": "../../../../../schema/connector_details.json",
+  "description": "../description.md",
+  "install_text": "The **WT201** is a LoRaWAN® thermostat designed for HVAC control, supporting systems such as heat pumps, AC units, boilers, furnaces, AHUs, and PTAC devices. It offers multiple modes: heat/cool, emergency heat, auto, and fan. The thermostat features a 4.2” LCD screen and an embedded temperature sensor, monitoring ambient temperature and humidity. It supports scheduled programming and remote management via a cloud system, with real-time monitoring for alerts on temperature changes. The device can store up to 1,000 data entries and integrates with hotel room card systems. It is compatible with the Milesight LoRaWAN® gateway and Milesight IoT Cloud solution, enabling remote control and D2D protocol.",
```
Comment (Freddyminu) [inline decoders/connector/milesight/wt201/v1.0.0/payload-config.jsonc:4]: Here.

**Review by Freddyminu — CHANGES_REQUESTED** [review]

File: decoders/connector/milesight/em400-mud-nb-iot/v1.0.0/payload.ts (hunk 8 above)
Comment (mateuscardosodeveloper) [reply decoders/connector/milesight/em400-mud-nb-iot/v1.0.0/payload.ts:76]: I'm creating the location variable, following the same interface on the @tago-io/sdk.

Line 204-211
```typescript
if (sensorData.latitude !== undefined && sensorData.longitude !== undefined) {
        parsedTagoObj.push({
          variable: "location",
          value: `${sensorData.latitude},${sensorData.longitude}`,
          location: { type: "Point", coordinates: [sensorData.latitude, sensorData.longitude] },
          group,
          time,
        });
      }
```

Comment (Freddyminu) [reply decoders/connector/milesight/em400-mud-nb-iot/v1.0.0/payload.ts:76]: Perfect, consider this resolved then.

**Review by Freddyminu — APPROVED** [review]

**Pull request merged**